proto/api_v3/query_service.proto

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax="proto3";
+
+package jaeger.api_v3;
+
+import "opentelemetry/proto/trace/v1/trace.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+
+option go_package = "api_v3";
+option java_package = "io.jaegertracing.api_v3";
+
+// Request object to get a trace.
+message GetTraceRequest {
+  // Hex encoded 128 bit trace_id
+  string trace_id = 1;
+}
+
+// A single response chunk holds a single trace.
+// Note that a subsequent request to get a trace might return a different result
+// because spans are reported in asynchronous chunks.
+message SpansResponseChunk {
+  // OTLP resource spans
+  // see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto
+  // In case of JSON format the ids (trace_id, span_id, parent_id) are encoded in base64 even though OpenTelemetry specification
+  // mandates to use hex encoding (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+  // Base64 is chosen to keep compatibility with JSONPb codec.
+  repeated opentelemetry.proto.trace.v1.ResourceSpans resource_spans = 1;
+}
+
+// Query parameters to find traces.
+message TraceQueryParameters {
+  string service_name = 1;
+  string operation_name = 2;
+  map<string, string> attributes = 3;
+  google.protobuf.Timestamp start_time_min = 4;
+  google.protobuf.Timestamp start_time_max = 5;
+  google.protobuf.Duration duration_min = 6;
+  google.protobuf.Duration duration_max = 7;
+  int32 num_traces = 8;
+}
+
+// Request object to search traces.
+message FindTracesRequest {
+  TraceQueryParameters query = 1;
+}
+
+// Request object to get service names.
+message GetServicesRequest {}
+
+// Response object to get service names.
+message GetServicesResponse {
+  repeated string services = 1;
+}
+
+// Request object to search for operation names.
+message GetOperationsRequest {
+  string service = 1;
+  string span_kind = 2;
+}
+
+// Operation encapsulates information about operation.
+message Operation {
+  string name = 1;
+  string span_kind = 2;
+}
+
+// Response object to get operation names.
+message GetOperationsResponse {
+  repeated Operation operations = 1;
+}
+
+service QueryService {
+  // GetTraces returns single trace.
+  // Note that the JSON response over HTTP is wrapped into result envelope "{"result": ...}"
+  // and so the response cannot be directly unmarshalled. This can be fixed by standard JSON
+  // parsing libraries or string manipulation to remove the envelope.
+  rpc GetTrace(GetTraceRequest) returns (stream SpansResponseChunk) {}
+
+  // GetTraces searches for traces.
```

Comment:
```suggestion
  // FindTraces searches for traces.
```

Reply:
The method is called `GetTraces`.

Here is the API mapping 

```
type: google.api.Service
config_version: 3
http:
  rules:
    - selector: jaeger.api_v3.QueryService.GetTrace
      get: /v3/traces/{trace_id}
    - selector: jaeger.api_v3.QueryService.GetTraces
      get: /v3/traces
    - selector: jaeger.api_v3.QueryService.GetServices
      get: /v3/services
    - selector: jaeger.api_v3.QueryService.GetOperations
      get: /v3/operations
```

I didn't use `FindTraces` on purpose to better match the usual REST semantics.

Reply:
* REST endpoints do not need to dictate how gRPC method is called. This will become the default gRPC API
* I would rather have consistency in the API across the backend, everywhere else (e.g. storage API) this is called FindTraces
* should we even have this method that returns full traces, or just have one that returns IDs?
* we do need a method like GetTraces that accepts a collection of trace IDs

Reply:
> should we even have this method that returns full traces, or just have one that returns IDs?